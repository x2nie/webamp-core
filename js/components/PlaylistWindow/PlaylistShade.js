import React from "react";
import { connect } from "react-redux";
import classnames from "classnames";
import { getTimeStr } from "../../utils";
import {
  TOGGLE_PLAYLIST_WINDOW,
  TOGGLE_PLAYLIST_SHADE_MODE,
  SET_FOCUSED_WINDOW
} from "../../actionTypes";

import {
  WINDOWS,
  PLAYLIST_RESIZE_SEGMENT_WIDTH,
  MIN_PLAYLIST_WINDOW_WIDTH
} from "../../constants";
import CharacterString from "../CharacterString";

import { getOrderedTracks } from "../../selectors";

const PlaylistShade = props => {
  const {
    toggleShade,
    close,
    focusPlaylist,
    playlistSize,
    focused,
    name,
    length
  } = props;

  const style = {
    width: `${MIN_PLAYLIST_WINDOW_WIDTH +
      playlistSize[0] * PLAYLIST_RESIZE_SEGMENT_WIDTH}px`
  };

  const classes = classnames("window", "draggable", {
    selected: focused === WINDOWS.PLAYLIST
  });

  return (
    <div
      id="playlist-window-shade"
      className={classes}
      style={{ width: style.width }}
      onMouseDown={focusPlaylist}
    >
      <div className="left">
        <div className="right draggable">
          <CharacterString id="playlist-shade-track-title">
            {name}
          </CharacterString>
          {/* TODO: Ellipisize */}
          <CharacterString id="playlist-shade-time">
            {getTimeStr(length)}
          </CharacterString>
          {/* TODO: Resize handle */}
          <div id="playlist-shade-button" onClick={toggleShade} />
          <div id="playlist-close-button" onClick={close} />
        </div>
      </div>
    </div>
  );
};

const mapDispatchToProps = dispatch => ({
  focusPlaylist: () =>
    dispatch({ type: SET_FOCUSED_WINDOW, window: WINDOWS.PLAYLIST }),
  close: () => dispatch({ type: TOGGLE_PLAYLIST_WINDOW }),
  toggleShade: () => dispatch({ type: TOGGLE_PLAYLIST_SHADE_MODE })
});

const mapStateToProps = state => {
  const {
    windows: { focused },
    display: { skinPlaylistStyle, playlistSize, playlistShade },
    media: { length, name }
  } = state;
  return {
    focused,
    skinPlaylistStyle,
    playlistSize,
    playlistShade,
    trackOrder: getOrderedTracks(state),
    length,
    name
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(PlaylistShade);
