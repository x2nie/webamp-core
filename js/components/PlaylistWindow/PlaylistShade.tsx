import React, { useMemo } from "react";
import { connect } from "react-redux";
import classnames from "classnames";
import { getTimeStr } from "../../utils";
import { SET_FOCUSED_WINDOW } from "../../actionTypes";
import * as Selectors from "../../selectors";

import {
  WINDOWS,
  WINDOW_RESIZE_SEGMENT_WIDTH,
  WINDOW_WIDTH,
  CHARACTER_WIDTH,
  UTF8_ELLIPSIS,
} from "../../constants";
import { togglePlaylistShadeMode, closeWindow } from "../../actionCreators";
import CharacterString from "../CharacterString";
import PlaylistResizeTarget from "./PlaylistResizeTarget";
import { AppState, WindowId, Dispatch } from "../../types";

interface StateProps {
  name: string | null;
  duration: number | null;
  playlistSize: [number, number];
  focused: WindowId | null;
  trackOrder: number[];
}

interface DispatchProps {
  focusPlaylist: () => void;
  close: () => void;
  toggleShade: () => void;
}

type Props = StateProps & DispatchProps;

function PlaylistShade({
  name,
  playlistSize,
  duration,
  toggleShade,
  close,
  focusPlaylist,
  focused,
}: Props) {
  const addedWidth = playlistSize[0] * WINDOW_RESIZE_SEGMENT_WIDTH;

  const trimmedName = useMemo(() => {
    if (name == null) {
      return "[No file]";
    }

    const MIN_NAME_WIDTH = 205;

    const nameLength = (MIN_NAME_WIDTH + addedWidth) / CHARACTER_WIDTH;
    return name.length > nameLength
      ? name.slice(0, nameLength - 1) + UTF8_ELLIPSIS
      : name;
  }, [name, addedWidth]);

  const time = useMemo(() => {
    return name == null ? "" : getTimeStr(duration);
  }, [name, duration]);

  return (
    <div
      id="playlist-window-shade"
      className={classnames("window", "draggable", {
        selected: focused === WINDOWS.PLAYLIST,
      })}
      style={{ width: `${WINDOW_WIDTH + addedWidth}px` }}
      onMouseDown={focusPlaylist}
      onDoubleClick={toggleShade}
    >
      <div className="left">
        <div className="right draggable">
          <div id="playlist-shade-track-title">
            <CharacterString>{trimmedName}</CharacterString>
          </div>
          <div id="playlist-shade-time">
            <CharacterString>{time}</CharacterString>
          </div>
          <PlaylistResizeTarget widthOnly />
          <div id="playlist-shade-button" onClick={toggleShade} />
          <div id="playlist-close-button" onClick={close} />
        </div>
      </div>
    </div>
  );
}

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => {
  return {
    focusPlaylist: () =>
      dispatch({
        type: SET_FOCUSED_WINDOW,
        window: WINDOWS.PLAYLIST,
      }),
    close: () => dispatch(closeWindow("playlist")),
    toggleShade: () => dispatch(togglePlaylistShadeMode()),
  };
};

const mapStateToProps = (state: AppState): StateProps => {
  const duration = Selectors.getDuration(state);
  const {
    windows: { focused },
  } = state;
  return {
    focused,
    playlistSize: Selectors.getWindowSize(state)("playlist"),
    trackOrder: Selectors.getOrderedTracks(state),
    duration,
    name: Selectors.getMinimalMediaText(state),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(PlaylistShade);
