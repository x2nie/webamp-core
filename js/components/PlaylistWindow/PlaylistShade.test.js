const media = {
  addEventListener: jest.fn(),
  _analyser: null
};

window.requestAnimationFrame = () => {};

import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import getStore from "../../store";

import PlaylistShade from "./PlaylistShade";

describe("PlaylistShade", () => {
  let store;
  beforeEach(() => {
    store = getStore(media);
  });

  it("renders to snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <PlaylistShade />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
