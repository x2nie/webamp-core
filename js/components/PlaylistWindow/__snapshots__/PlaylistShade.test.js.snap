// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistShade renders to snapshot 1`] = `
<div
  className="window draggable"
  id="playlist-window-shade"
  onMouseDown={[Function]}
  style={
    Object {
      "width": "275px",
    }
  }
>
  <div
    className="left"
  >
    <div
      className="right draggable"
    >
      <div
        id="playlist-shade-track-title"
      />
      <div
        id="playlist-shade-time"
      >
        <div
          className=" character character-48"
        >
          0
        </div>
        <div
          className=" character character-48"
        >
          0
        </div>
        <div
          className=" character character-58"
        >
          :
        </div>
        <div
          className=" character character-48"
        >
          0
        </div>
        <div
          className=" character character-48"
        >
          0
        </div>
      </div>
      <div
        id="playlist-shade-button"
        onClick={[Function]}
      />
      <div
        id="playlist-close-button"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
